cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java

```diff
@@ -848,4 +883,117 @@ private void maybeReexecuteTasks() {
       }
     }
   }
+
+  /**
+   * We finish execution after Step. 2: moving leadership. Checking pending leadership or partition movement ascertains
+   * that execution is really done (instead of quitting due to Exception).
+   * @return true if execution for proposal is done.
+   */
+  private boolean executionSucceeded() {
+    return _executorState.state() == LEADER_MOVEMENT_TASK_IN_PROGRESS &&
```

Comment:
should here be `NO_TASK_IN_PROGRESS`?

Reply:
The state is not set to `NO_TASK_IN_PROGRESS` until we call `clearCompletedExecution()`, `clearCompletedExecution()` also clears `_uuid` which we need. I think the conditions in `executionSucceeded()` is also valid.

Reply:
I see, then what about `REPLICA_MOVEMENT_TASK_IN_PROGRESS`? If there is replica movement task, shouldn't the state be parked at `REPLICA_MOVEMENT_TASK_IN_PROGRESS`?

Reply:
I see, then what about REPLICA_MOVEMENT_TASK_IN_PROGRESS? If there is replica movement task, shouldn't the state be parked at REPLICA_MOVEMENT_TASK_IN_PROGRESS?

Reply:
the executor goes from `Starting_execution` to `REPLICA_MOVEMENT_TASK_IN_PROGRESS` to  `LEADER_MOVEMENT_TASK_IN_PROGRESS` to finishing, in this sequence, so we always go through `LEADER_MOVEMENT_TASK_IN_PROGRESS`.

Reply:
looks like it could park at `STOPPING_EXECUTION` state if it is stopped by user or CC?(see [code](https://github.com/linkedin/cruise-control/blob/2b98cdbe4380159b912dcbcc5a4e3082f52c1843/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java#L557))

Reply:
@efeg  We have a discussion offline, the conclusion is this is expected.  if execution is parked at `STOPPING_EXECUTION`, the execution is not considered successful.

Reply:
@efeg We discussed Kun's case. Basically, execution state will always go through `REPLICA_MOVEMENT_TASK_IN_PROGRESS` and `LEADER_MOVEMENT_TASK_IN_PROGRESS`, at which point, if the execution finished, the remaining leadership and replica movement is 0. If the task reaches this point, we consider it succeeded. All other paths are considered failed or stopped. Could you provide an example where the condition gives false positive?
I can move it as you suggested, although `_executorNotifier.sendNotification` is a separate action from `clearCompletedExecution` and it feels cleaner to keep them separate.